src/libutil/archive.cc

```diff
@@ -36,91 +37,84 @@ static GlobalConfig::Register rArchiveSettings(&archiveSettings);
 PathFilter defaultPathFilter = [](const Path &) { return true; };
 
 
-static void dumpContents(const Path & path, off_t size,
-    Sink & sink)
+void SourceAccessor::dumpPath(
+    const CanonPath & path,
+    Sink & sink,
+    PathFilter & filter)
 {
-    sink << "contents" << size;
-
-    AutoCloseFD fd = open(path.c_str(), O_RDONLY | O_CLOEXEC);
-    if (!fd) throw SysError("opening file '%1%'", path);
+    auto dumpContents = [&](const CanonPath & path)
+    {
+        /* It would be nice if this was streaming, but we need the
+           size before the contents. */
+        auto s = readFile(path);
```

Comment:
This looks like a pretty serious regression. We worked hard to get rid of arbitrary large buffers in Nix; we shouldn't go back.

Reply:
`readFile()` now supports reading into a sink.